File: src/contests/top150/n238_product_of_array_except_self/Solution.java
package contests.top150.n238_product_of_array_except_self;

import java.util.Arrays;

public class Solution {

    public int[] productExceptSelf(int[] nums) {
        int[] arr = new int[nums.length];
        int left = 1, right = 1;

        for (int i = 0; i < nums.length; i++) {
            arr[i] = left;
            left *= arr[i];
        }

        for (int i = nums.length - 1; i >= 0; i--) {
            arr[i] *= right;
            right *= arr[i];
        }

        return arr;
    }

    public static void main(String[] args) {
        var s = new Solution();
        System.out.println(Arrays.toString(s.productExceptSelf(new int[]{1, 2, 3, 4})));
    }

}

File: src/contests/top150/n238_product_of_array_except_self/solution.go
package main

func productExceptSelf(nums []int) []int {
	res := make([]int, len(nums))
	l := 1
	r := 1

	for i := 0; i < len(nums); i++ {
		res[i] = l
		l *= nums[i]
	}

	for i := len(nums) - 1; i >= 0; i-- {
		res[i] *= r
		r *= nums[i]
	}
	return res
}

func main() {
	println(productExceptSelf([]int{1, 2, 3, 4}))
}
